mydemoproject/demoapp1/views.py

```diff
@@ -1,17 +1,186 @@
-from django.shortcuts import render
-from django.http import HttpResponse
+from django.contrib import messages
+from django.shortcuts import redirect
+from django.views.generic import (ListView, DeleteView,
+                                  UpdateView, CreateView,
+                                  TemplateView)
+from django.urls import reverse_lazy
+from .models.book import Book
+from .models.inventory import Inventory
+from .models.user import CustomUser
+from .models.borrowed_book import BorrowedBook
 
-# Create your views here.
 
-def show_books(request):
-    return render(request, 'demoapp1/books.html')
+# ListView for Users
+class UserListView(ListView):
+    model = CustomUser
+    template_name = 'demoapp1/users.html'
+    context_object_name = 'users'
 
-def show_users(request):
-    return render(request, 'demoapp1/users.html')
 
+# DeleteView for Users
+class UserDeleteView(DeleteView):
+    model = CustomUser
+    success_url = reverse_lazy('show_user')
+    pk_url_kwarg = 'user_id'
 
-def show_inventory(request):
-    return render(request, 'demoapp1/inventory.html')
 
-def show_borrowed(request):     
-    return render(request, 'demoapp1/borrowed.html')
+# UpdateView for User
+class UserUpdateView(UpdateView):
+    model = CustomUser
+    fields = ['username', 'email']
+    template_name = 'demoapp1/edit_form.html'
+    context_object_name = 'user'
+    success_url = reverse_lazy('show_user')
+    pk_url_kwarg = 'user_id'
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context['object_type'] = 'User'
+        context['edit_url'] = 'edit_user'
+        return context
+
+
+# ListView for Books
+class BookListView(ListView):
+    model = Book
+    template_name = 'demoapp1/books.html'
+    context_object_name = 'books'
+
+    def get_queryset(self):
+        return Book.objects.select_related('inventory')
+
+
+# CreateView for User
+class BookCreateView(CreateView):
+    model = Book
+    fields = ['book_name', 'author_name']
+    template_name = 'demoapp1/books.html'
+    context_object_name = 'book'
+    success_url = reverse_lazy('show_book')
+
+    def form_valid(self, form):
```

Comment:
When is this method invoked?

Reply:
only invoked when al the form validations has been succesfully validated